interp/api.go

```diff
@@ -329,14 +332,37 @@ func CallHandler(f CallHandlerFunc) RunnerOption {
 	}
 }
 
-// ExecHandler sets the command execution handler. See ExecHandlerFunc for more info.
+// ExecHandler sets one command execution handler,
+// which replaces DefaultExecHandler(2 * time.Second).
+//
+// Deprecated: use ExecHandlers instead, which allows for middleware handlers.
 func ExecHandler(f ExecHandlerFunc) RunnerOption {
 	return func(r *Runner) error {
 		r.execHandler = f
 		return nil
 	}
 }
 
+// ExecHandlers appends command execution handler middlewares.
+// Each middleware is run in order from first to last,
+// and chooses whether to fall back to the next middleware.
```

Comment:
This sentence is contradictory to me.  Are the middlewares run in order first to last, or does each one choose whether to fall back to the next one?

The api seems awkward, too, since each middleware has to correctly call the `next` function.  It seems like it'd be better if the interpreter looped over the middlewares itself, and each one should return either a `result, continue` pair (either one or the other but not both), or better yet a single result with a special `interp.CallNextExecHandler` (or something) value to signal the scaffolding to call the next handler.

Reply:
You're right that the godoc is a bit awkward. How about:

> The middlewares are chained from first to last, and the first is called by the runner.
> The middleware can choose whether to handle the operation and return an error,
> or to not handle the operation and fall back to calling the `next` middleware.

About the awkward API - I actually implemented precisely what you mean in the original coreutils branch. See: https://github.com/mvdan/sh/compare/93-coreutils

It is conceptually simpler, as there are no `func(next Handler) Handler` shenanigans which may confuse the user. However, it is also less powerful. When a middleware can only handle an operation or return "skip me", those are its only two options. With the "next" call approach, you can imagine middlewares which _alter the operation parameters_ while letting the next middleware handle the operation.

For example, imagine an exec middleware which wants to replace all calls to `chrome` with `firefox`. With the "skip me" approach, it would have to detect `args[0] == "chrome"` and then fully implement the process execution, e.g. by calling `DefaultExecHandler` with the altered `args[0]`. If there was a middleware right after which would also have altered how this process execution happened, by e.g. using a higher exec timeout, that would be ignored.

With the "next" approach, the logic can just be as follows, and the middlewares coming later in the chain are obeyed as usual:

```
if args[0] == "chrome" {
    args[0] = "firefox"
}
next(args)
```

This is heavily used in go-chi to alter requests or responses without actually _finishing_ them. For example, see https://github.com/go-chi/chi/blob/v5.0.8/middleware/clean_path.go#L12, which cleans the path of an HTTP request and lets the next middleware do its thing. I think that's very powerful.

Reply:
The godoc you suggested is better.  It leaves out the nuance that a) you could filter/replace/log the args when calling next, and b) you could filter/log the return value when next returns.  If you feel like that's implicitly obvious, no problem.  If somebody's familiar with this kind of API (obviously I wasn't), then it probably would be obvious, but if they're not, it might not be.

You're right, this api is more powerful than just "handle or skip".  I was going to suggest that you go with my suggested api and just pass a pointer to args, which would let you change them, but then that wouldn't let you similarly filter (or even just log) the return value.

Reply:
That's a good point about other options, I'll expand it a bit further.

One disadvantage of this more powerful API is that being able to call "next" more than once, or with entirely wrong parameters, is a potential footgun. But hopefully that kind of mistake will be pretty easy to spot.

Reply:
I've expanded the godoc.